### PR TITLE
 add CI to compose supergraph schema and publish to OCI

### PR DESCRIPTION
**Pull request by iamvigneshwars**

### CODE REVIEW

File: schema/initial.graphql
Comment (garryod) [inline schema/initial.graphql:1]: Probably best calling this `placeholder`

File: .github/workflows/schema-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    # Deduplicate jobs from pull requests and branch pushes within the same repo.
+    if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name != github.repository
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout source
+        uses: actions/checkout@v4.1.4
+
+      - name: Install Rover CLI
+        run: |
+          curl -sSL https://rover.apollo.dev/nix/v0.23.0-rc.3 | sh
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:18]: How are we going to keep this updated?

File: .github/workflows/schema-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          echo "$HOME/.rover/bin" >> $GITHUB_PATH
+
+      - name: Compose Schema
+        run: |
+          rover supergraph compose --config schema/supergraph-config.yaml --elv2-license=accept > supergraph.graphql
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:23]: Just makes this a bit more readable
```suggestion
      - name: Compose Schema
        run: >
          rover supergraph compose
          --config schema/supergraph-config.yaml
          --elv2-license=accept
          > supergraph.graphql
```

File: .github/workflows/schema-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          cat <<EOF > ${{ matrix.chart}}/templates/supergraph-configmap.yaml
+          apiVersion: v1
+          kind: ConfigMap
+          metadata:
+            name: supergraph-config
+          data:
+            supergraph.graphql: |
+          EOF
+
+          sed 's/^/    /' supergraph.graphql >> ${{ matrix.chart }}/templates/supergraph-configmap.yaml
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:75]: Could this be made more concise with a template `ConfigMap` and `yq`?

File: .github/workflows/schema-compose.yaml [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    needs:
+      - generate
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+    strategy:
+      matrix:
+        chart:
+          - supergraph
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:44]: What other `chart` entries are we expecting?

File: .github/workflows/schema-compose.yaml
Comment (garryod) [inline .github/workflows/schema-compose.yaml:1]: We should probably label all images with `latest` and those which are built on a tagged release with `stable`, so we can use those tags in ArgoCD

**Review by garryod — CHANGES_REQUESTED** [review]

File: .github/workflows/schema-compose.yaml (hunk 1 above)
Comment (iamvigneshwars) [reply .github/workflows/schema-compose.yaml:18]: Manually I guess? I found a github action in market place but its using older version I guess. Any way we only using it to compose supergraphs, should have any issues unless there's a major updates?

File: .github/workflows/schema-compose.yaml (hunk 4 above)
Comment (iamvigneshwars) [reply .github/workflows/schema-compose.yaml:44]: None for now!

Comment (garryod) [reply .github/workflows/schema-compose.yaml:44]: Would it be worth getting rid of the `matrix` then? Just to keep things simple

Comment (iamvigneshwars) [reply .github/workflows/schema-compose.yaml:44]: Yeah, I got rid of that, that's why I resolved the conversation.

File: .github/workflows/schema-compose.yaml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        if: github.event_name == 'push' && github.ref == 'refs/heads/main'
+        run: helm push $(ls supergraph-*.tgz) ${{ env.IMAGE_REPOSITORY }}
+
+  compare_schema:
+    runs-on: ubuntu-latest
+    if: ${{ github.event_name == 'pull_request' }}
+    steps:
+      - name: Install Rover CLI
+        run: |
+          curl -sSL https://rover.apollo.dev/nix/v0.23.0-rc.3 | sh
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:80]: Can we use `https://rover.apollo.dev/nix/latest` here too?

File: .github/workflows/schema-compose.yaml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+          --config supergraph-config.yaml
+          --elv2-license=accept
+          > supergraph.graphql
+
+      - name: Generate Schema Diff
+        run: |
+          git --no-pager diff --no-index --no-exit-code --minimal /tmp/old_supergraph.graphql supergraph.graphql > diff.md
+          if [ -s diff.md ]; then echo "has_diff=true" >> $GITHUB_ENV; fi
+          sed -i 1,4d diff.md
+          sed -i '1s/^/```diff\n/; $a```' diff.md
```
Comment (garryod) [inline .github/workflows/schema-compose.yaml:110]: Think the `sed`s will fail if there is no diff, should probably put them inside the `if`

**Review by garryod — CHANGES_REQUESTED** [review]

File: .github/workflows/schema-compose.yaml (hunk 6 above)
Comment (garryod) [reply .github/workflows/schema-compose.yaml:110]: Discussed offline, this PR doesn't have any diff :facepalm:

**Review by garryod — APPROVED** [review]
LGTM

**Pull request merged**